### PR TITLE
#3505 Add date modification parsing

### PR DESCRIPTION
**Pull request by dcechano**
This commit addresses [#3505](https://github.com/uutils/coreutils/issues/3505). I added a `parse_date_modification.rs` module to `uucore::parser` and used it in `touch.rs`. I added tests to `touch.rs` to prove the parser works and is not buggy. In theory these changes can also help with similar issues in `date.rs` but I wanted to keep the PR to one issue at a time.

Thank you for considering my pull request and I welcome feedback if you guys have any.

### CODE REVIEW

File: src/uu/touch/src/touch.rs [hunk 1]
```diff
@@ -339,8 +341,14 @@ fn parse_date(ref_time: DateTime<Local>, s: &str) -> UResult<FileTime> {
...
-    if let Ok(parsed) = NaiveDateTime::parse_from_str(s, format::POSIX_LOCALE) {
-        return Ok(datetime_to_filetime(&parsed.and_utc()));
+    if let Ok((parsed, modifier)) = NaiveDateTime::parse_and_remainder(s, format::POSIX_LOCALE) {
+        if modifier.is_empty() {
+            return Ok(datetime_to_filetime(&parsed.and_utc()));
+        }
+        return match date_from_modifier(modifier, parsed) {
+            Ok(new_date) => Ok(datetime_to_filetime(&new_date.and_utc())),
+            Err(e) => Err(e),
+        };
```
Comment (sylvestre) [inline src/uu/touch/src/touch.rs:351]: I think it could be simplified with:
```suggestion
return date_from_modifier(modifier, parsed).map(|new_date| datetime_to_filetime(&new_date.and_utc()));
```

File: src/uu/touch/src/touch.rs [hunk 2]
```diff
@@ -352,18 +360,30 @@ fn parse_date(ref_time: DateTime<Local>, s: &str) -> UResult<FileTime> {
...
-        if let Ok(parsed) = NaiveDateTime::parse_from_str(s, fmt) {
-            return Ok(datetime_to_filetime(&parsed.and_utc()));
+        if let Ok((parsed, modifier)) = NaiveDateTime::parse_and_remainder(s, fmt) {
+            if modifier.is_empty() {
+                return Ok(datetime_to_filetime(&parsed.and_utc()));
+            }
+            return match date_from_modifier(modifier, parsed) {
+                Ok(new_date) => Ok(datetime_to_filetime(&new_date.and_utc())),
+                Err(e) => Err(e),
+            };
```
Comment (sylvestre) [inline src/uu/touch/src/touch.rs:370]: can probably be simplified too by something like
```suggestion
if let Ok((parsed, modifier)) = NaiveDateTime::parse_and_remainder(s, fmt) {
    return if modifier.is_empty() {
        Ok(datetime_to_filetime(&parsed.and_utc()))
    } else {
        date_from_modifier(modifier, parsed).map(|new_date| datetime_to_filetime(&new_date.and_utc()))
    };
}

```

File: src/uu/touch/src/touch.rs [hunk 3]
```diff
@@ -352,18 +360,30 @@ fn parse_date(ref_time: DateTime<Local>, s: &str) -> UResult<FileTime> {
...
-    if let Ok(parsed_date) = NaiveDate::parse_from_str(s, format::ISO_8601) {
+    if let Ok((parsed_date, modifier)) = NaiveDate::parse_and_remainder(s, format::ISO_8601) {
         let parsed = Local
             .from_local_datetime(&parsed_date.and_time(NaiveTime::MIN))
             .unwrap();
-        return Ok(datetime_to_filetime(&parsed));
+        if modifier.is_empty() {
+            return Ok(datetime_to_filetime(&parsed));
+        }
+        return match date_from_modifier(modifier, parsed) {
```
Comment (sylvestre) [inline src/uu/touch/src/touch.rs:383]: same, Err(e) => Err(e),
can often be simplified

File: src/uucore/src/lib/parser/parse_date_modifier.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
+use std::collections::HashMap;
```
Comment (sylvestre) [inline src/uucore/src/lib/parser/parse_date_modifier.rs:1]: please add a license header here

**Comment by sylvestre** [comment]
Lot of work probably went into this, bravo
please add integration tests here too:
https://github.com/uutils/coreutils/blob/main/tests/by-util/test_touch.rs

File: src/uu/touch/src/touch.rs [hunk 5]
```diff
@@ -437,6 +457,96 @@ fn parse_timestamp(s: &str) -> UResult<FileTime> {
...
+                match chrono {
+                    ChronoUnit::Year => {
+                        if time > (i64::MAX / 12) {
+                            return Err(USimpleError::new(
+                                1,
+                                format!("Unable to parse modifier: {modifier}"),
+                            ));
+                        }
+                        date = if time >= 0 {
+                            date.add(Months::new((12 * time) as u32))
```
Comment (sylvestre) [inline src/uu/touch/src/touch.rs:488]: codecov suggests that this line isn't covered, could you please fix it? thanks

File: src/uucore/src/lib/parser/parse_date_modifier.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        }
+        let bytes = &self.haystack[self.cursor..];
+        if bytes[0] == b'+' || bytes[0] == b'-' || (bytes[0] > 47 && bytes[0] <= 57) {
+            let mut nums = vec![bytes[0] as char];
+            let mut i = 1;
+            loop {
+                if i >= bytes.len() {
+                    break;
+                }
+                if let n @ 48..=57 = bytes[i] {
```
Comment (sylvestre) [inline src/uucore/src/lib/parser/parse_date_modifier.rs:97]: please replace 48 & 57 by variable names

File: src/uucore/src/lib/parser/parse_date_modifier.rs [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+                        }
+                    }
+                }
+            }
+        }
+        Ok(map)
+    }
+
+    fn parse_num(&mut self) -> Result<i64, ParseError> {
+        while self.cursor < self.haystack.len() && self.haystack[self.cursor].is_ascii_whitespace()
```
Comment (sylvestre) [inline src/uucore/src/lib/parser/parse_date_modifier.rs:82]: please create a function:

```
fn skip_whitespace(&mut self) {
    while self.cursor < self.haystack.len() && self.haystack[self.cursor].is_ascii_whitespace() {
        self.cursor += 1;
    }
}

```

File: src/uucore/src/lib/parser/parse_date_modifier.rs [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+            }
+            let n_as_string = nums.iter().collect::<String>();
+            n_as_string.parse::<i64>().map_err(|_| IntOverflow)
+        } else {
+            Err(UnexpectedToken)
+        }
+    }
+
+    fn parse_unit(&mut self) -> Result<ChronoUnit, ParseError> {
+        while self.cursor < self.haystack.len() && self.haystack[self.cursor].is_ascii_whitespace()
```
Comment (sylvestre) [inline src/uucore/src/lib/parser/parse_date_modifier.rs:122]: and replace this one too

File: src/uucore/src/lib/parser/parse_date_modifier.rs [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        {
+            self.cursor += 1;
+        }
+        if self.cursor >= self.haystack.len() {
+            return Err(NoUnitPresent);
+        }
+        let bytes = &self.haystack[self.cursor..].to_ascii_lowercase();
+        match bytes[0] {
+            b'd' => {
```
Comment (sylvestre) [inline src/uucore/src/lib/parser/parse_date_modifier.rs:131]: i can probably by simplified by something like:

```
    let units = [
        (b'd', "day", ChronoUnit::Day),
        (b'w', "week", ChronoUnit::Week),
        (b'm', "month", ChronoUnit::Month),
        (b'y', "year", ChronoUnit::Year),
        (b'h', "hour", ChronoUnit::Hour),
        (b's', "second", ChronoUnit::Second),

    ];
    for &(byte, unit_str, chrono_unit) in &units {
        if bytes.starts_with(unit_str) {
            self.cursor += unit_str.len();
            return Ok(chrono_unit);
        }
    }
```

**Review by tertsdiepraam — COMMENTED** [review]
@sylvestre, you seem to have a pretty good grasp on this already. Shouldn't this be fixed in `parse_datetime`?

**Comment by dcechano** [comment]
> @sylvestre, you seem to have a pretty good grasp on this already. Shouldn't this be fixed in `parse_datetime`?

I actually did consider this, but I figured it is not technically parsing a date but a modifier to a date. I can move the changes over to `parse_datetime` if needed.

File: src/uucore/src/lib/parser/parse_date_modifier.rs (hunk 9 above)
Comment (dcechano) [reply src/uucore/src/lib/parser/parse_date_modifier.rs:131]: Thank you for reviewing this PR I can implement the changes later this evening.

**Comment by tertsdiepraam** [comment]
Yes, that's what parse datetime is for. It's supposed to parse the format described here: https://www.gnu.org/software/coreutils/manual/html_node/Date-input-formats.html, which includes this part I believe. Porting it would be great!

**Comment by dcechano** [comment]
I made the suggested changes including adding integration testing. I also increased testing to satisfy the code coverage bot, however I'm not sure I was able to get everything. I apologize in advance if the I did something incorrectly; I don't know git very well.

**Comment by sylvestre** [comment]
what do you think about moving it into https://github.com/uutils/parse_datetime/ ?

**Comment by dcechano** [comment]
> what do you think about moving it into https://github.com/uutils/parse_datetime/ ?


Oh, I see. I misunderstood @tertsdiepraam. Yes, I'm willing to move my work to the [`parse_datetime`](https://github.com/uutils/parse_datetime/). However, I may need more time to investigate the best way to integrate them smoothly. 

One observation is that `parse_datetime` uses [`Regex`](https://github.com/rust-lang/regex) for parsing which will render some of my algorithm redundant. Another is that `parse_datetime` already appears to do something similar but evidently is not parsing `'2022-05-15 +01 Month'`, as pointed out in issue #3505.

In short, I just need to a day or two to reorient myself. But I am confident I can figure it out since I already gained a decent understanding while working on this PR.

**Comment by dcechano** [comment]
@sylvestre @tertsdiepraam I ran into some interesting issues while working to transfer this PR to `parse_datetime`. I have opened an [issue](https://github.com/uutils/parse_datetime/issues/56) that I think warrants your inputs. Your guidance would be invaluable.

**Comment by sylvestre** [comment]
@dcechano sorry but do you have an update ? thanks

**Comment by dcechano** [comment]
> @dcechano sorry but do you have an update ? thanks

Yes sir. I am very sorry for the long delay but working out a solution that covers edge cases turned out to be quite challenging. I have been working in the [parse_datetime](https://github.com/uutils/parse_datetime) repo as suggested and getting a PR ready. 

What has been taking so long is that I had to implement the core logic (which was challenging in its own right) then integrate it into the larger library via `lib.rs`. That took some unanticipated refactoring, discovering bugs/edge cases ect. I had to find a way to take arbitrary strings with arbitrary formats and figure out how to hand off the appropriate part to `chrono` and take the remaining and process it via `parse-datetime`. This was quite a battle to get right. But I am very happy with the result. 

After the PR, which I have set a hard goal of getting in by the end of this week, `parse_datetime` will be able to take arbitrary input like `08-31-2022 12:30:00 +1 month` and spit out the result. The only restriction is that the time format has to be in a currently supported format.
I am just cleaning up a bit to make it presentable and implementing more tests to make sure that the code I am submitting is sufficiently robust.

**Comment by dcechano** [comment]
@sylvestre & @tertsdiepraam I have a [PR](https://github.com/uutils/parse_datetime/pull/57) in that is a major step in addressing #3505. I'm looking forward to hearing what you guys think. Thank you!

**Comment by sylvestre** [comment]
@dcechano ping ? :)

**Comment by RenjiSann** [comment]
The issue seems to be fixed anyway
